pkg/registry/extensions/daemonset/strategy.go

```diff
@@ -70,6 +70,15 @@ func (daemonSetStrategy) PrepareForUpdate(ctx genericapirequest.Context, obj, ol
 	// update is not allowed to set TemplateGeneration
 	newDaemonSet.Spec.TemplateGeneration = oldDaemonSet.Spec.TemplateGeneration
 
+	// Update is not allowed to set Spec.Selector for all groups/versions except extensions/v1beta1.
+	// If RequestInfo is nil, it is better to revert to old behavior (i.e. allow update to set Spec.Selector)
+	// to prevent unintentionally breaking users who may rely on the old behavior.
+	// TODO(#50791): after v1beta1 is retired, move selector immutability check to validation codes
+	requestInfo, found := genericapirequest.RequestInfoFrom(ctx)
+	if found && !(requestInfo.APIGroup == "extensions" && requestInfo.APIVersion == "v1beta1") {
```

Comment:
> when v1 is added, you really want this to require a decision to be made about this behavior

If there's a way to require this decision at compile time, I am definitely in favor. We can't return an error because the function signature doesn't allow it, so our only other option would be to panic. Then we would be relying on test coverage to check for mutating selectors in the new version.

If we have to rely on test coverage anyway, I would rather rely on tests being added at the point when this check moves to validation (after all legacy versions are removed). When we add the test to ensure the new validation triggers an error, that test will fail until we rip out this code.